src/logsoftmax.jl

```diff
@@ -0,0 +1,34 @@
+using Base.Threads
+
+function logsoftmax!(out::AbstractVecOrMat, xs::AbstractVecOrMat)
+  @threads for j = 1:size(xs, 2)
+    @inbounds begin
+      xi_max = xs[1, j]
+      for i = 1:size(out, 1)
+        xi_max = max(xi_max, xs[i, j])
+      end
+      s = zero(eltype(out))
+      for i = 1:size(out, 1)
+        s += exp(xs[i, j] - xi_max)
+      end
+      for i = 1:size(out, 1)
+        out[i, j] = xs[i, j] - log(s) - xi_max
```

Comment:
Do you remember why you subtract `xi_max` here?
IIRC the definition of logsoftmax normally is

```
x_max = maximum(xs)
s = log(sum(exp.(xs - x_max)))
out = xs .- s
```


```

Reply:
I think we are doing the same thing as pytorch. If you replace `s` with `log(sum(exp.(xs - x_max))) + x_max` you will get 
```
out = xs .- log(sum(exp.(xs - x_max))) - x_max
```